sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/DataSource.scala

```diff
@@ -132,13 +132,24 @@ case class DataSource(
       }.toArray
       new InMemoryFileIndex(sparkSession, globbedPaths, options, None)
     }
+
+    val dataSchema = userSpecifiedSchema.orElse {
```

Comment:
I would still keep this below the `if (justPartitioning)` area, because otherwise everytime someone performs a `df.mode("append").saveAsTable()` we will perform schema inference.